public/js/table-of-contents.js

```diff
@@ -0,0 +1,104 @@
+const MAX_TRIES = 20;
+
+/**
+ * Wait for an element to exist within the DOM, return that element
+ * Will reject if the attempts exceeds MAX_TRIES
+ * @param {string} selector
+ * @returns HTMLElement
+ */
+const waitForElement = (selector) =>
+  new Promise((resolve, reject) => {
+    let interval;
+    let tries = 0;
+
+    interval = setInterval(() => {
+      const element = document.body.querySelector(selector);
+      if (!element || element.length === 0) {
+        tries++;
+        if (tries > MAX_TRIES) {
+          clearInterval(interval);
+          reject('Could not find element');
+        }
+
+        return;
+      }
+
+      clearInterval(interval);
+      resolve(element);
+    }, 100);
+  });
+
+const createToc = async (config) => {
+  const tocElement = config.tocElement;
+  const titleElements = config.titleElements;
+
+  let tocElementDiv;
+  try {
+    tocElementDiv = await waitForElement(tocElement);
+  } catch (error) {
+    console.error(error);
+    return;
+  }
+
+  let tocUl = document.createElement('ul');
+  tocUl.id = 'list-toc-generated';
+  tocElementDiv.appendChild(tocUl);
+
+  // add class to all title elements
+  let tocElementNbr = 0;
+  for (let i = 0; i < titleElements.length; i++) {
+    let titleHierarchy = i + 1;
+    let titleElement = document.body.querySelectorAll(titleElements[i]);
+
+    titleElement.forEach(function (element) {
+      // add classes to the element
+      element.classList.add('title-element');
+      element.setAttribute('data-title-level', titleHierarchy);
+
+      // add id if doesn't exist
+      tocElementNbr++;
+      const idElement = element.id;
+      if (idElement == '') {
+        element.id = 'title-element-' + tocElementNbr;
```

Comment:
Template literals have had wide browser support for many years and is preferred over manual string concatenation, but of course not a blocker.

Reply:
I agree, this script should be made better. This thing was copied over from `pdd-web` though, and I don't really want to put any more work into it since I think we will (hopefully) move to calculate the TOC within React instead of a script.